browser-extension/mv3/src/service-worker/services/requestProcessor.ts

```diff
@@ -0,0 +1,177 @@
+      },
+    });
+  };
+
+  private hasIgnoredHeadersInRequest = (requestHeaders: Record<string, string>) => {
+    return IGNORED_HEADERS_ON_REDIRECT.some((header) => requestHeaders[header] || requestHeaders[header.toLowerCase()]);
+  };
+
+  private findMatchingRule = (rules: Rule[], url: string): Rule | undefined => {
+    return rules.find((rule) => rule.pairs.some((pair) => matchSourceUrl(pair.source, url)));
+  };
+
+  private createActionDetails = (
+    matchedRule: Rule,
+    requestDetails: AJAXRequestDetails,
+    actionType: ActionType
+  ): ActionDetails => {
+    switch (actionType) {
+      case ActionType.FORWARD_IGNORED_HEADERS:
+        return {
+          type: ActionType.FORWARD_IGNORED_HEADERS,
+          ignoredHeadersValue: IGNORED_HEADERS_ON_REDIRECT.map((header) => ({
+            name: header,
+            value: requestDetails.requestHeaders[header] || requestDetails.requestHeaders[header.toLowerCase()],
+          })),
+          redirectedUrl: this.getRedirectedUrl(requestDetails.url, matchedRule),
+        };
+    }
+  };
+
+  private getRedirectedUrl = (requestUrl: string, matchedRule: Rule) => {
+    const pair = matchedRule.pairs.find((pair) => matchSourceUrl(pair.source, requestUrl));
+    const redirectedUrl = pair.destination || requestUrl.replace(pair.from, pair.to);
+    if (redirectedUrl === requestUrl) {
+      return null;
+    }
+    return redirectedUrl;
+  };
+
+  private updateRequestSpecificRules = async (
+    tabId: number,
+    requestUrl: string,
+    ruleDetails: {
+      action: chrome.declarativeNetRequest.RuleAction;
+      condition: chrome.declarativeNetRequest.RuleCondition;
+    }
+  ) => {
+    let ruleId = parseInt(`${Date.now() % 1000000}${Math.floor(Math.random() * 1000)}`);
+
+    const sessionRulesMap = tabService.getData(tabId, TAB_SERVICE_DATA.SESSION_RULES_MAP);
+
+    let removeRuleIds = [];
+
+    if (sessionRulesMap?.[requestUrl]) {
+      ruleId = sessionRulesMap[requestUrl];
+      removeRuleIds.push(ruleId);
+    }
+
+    tabService.setData(tabId, TAB_SERVICE_DATA.SESSION_RULES_MAP, {
+      ...sessionRulesMap,
+      [requestUrl]: ruleId,
+    });
+
+    return updateSessionRules({
+      addRules: [
+        {
+          id: ruleId,
+          ...ruleDetails,
+        },
+      ],
+      removeRuleIds,
+    });
+  };
+
+  onBeforeAJAXRequest = async (tabId: number, requestDetails: AJAXRequestDetails) => {
+    if (Object.values(this.cachedRules).every((rules) => rules.length === 0)) {
+      return;
+    }
+
+    console.time("processRequest");
+    const requestActionDetails = await this.processRequest(requestDetails);
+    console.timeEnd("processRequest");
+    if (!requestActionDetails) {
+      return;
+    }
+
+    switch (requestActionDetails.type) {
+      case ActionType.FORWARD_IGNORED_HEADERS:
+        console.time("forwardIgnoredHeaders");
+        await this.forwardIgnoredHeaders(tabId, requestDetails, requestActionDetails);
```

Comment:
Going forward, there could be multiple actions to be applied in `onBeforeAJAXRequest`.
Instead of creating an action and then executing it, can we simply delegate this to a different function?
```js
public async onBeforeAJAXRequest (tabId: number, requestDetails: AJAXRequestDetails) {
  this.forwardIgnoredHeadersOnRedirect(tabId, requestDetails);
  this.handlePredefinedFunctions(tabId, requestDetails);
  // and so on...
}

private forwardIgnoredHeadersOnRedirect (tabId: number, requestDetails: AJAXRequestDetails) {
  if (!this.cachedRules.redirectRules?.length || !this.cachedRules.replaceRules?.length) {
    return;
  }

  if (!this.hasIgnoredHeadersInRequest(requestDetails.requestHeaders)) {
    return;
  }

  // check if matches any redirect/replace rule
  // apply new DNR rules
}


```